server/src/main/java/org/elasticsearch/rest/ChunkedRestResponseBody.java

```diff
@@ -92,20 +93,33 @@ public boolean isDone() {
 
             @Override
             public ReleasableBytesReference encodeChunk(int sizeHint, Recycler<BytesRef> recycler) throws IOException {
-                final RecyclerBytesStreamOutput chunkStream = new RecyclerBytesStreamOutput(recycler);
-                assert this.target == null;
-                this.target = chunkStream;
-                while (serialization.hasNext()) {
-                    serialization.next().toXContent(builder, params);
-                    if (chunkStream.size() >= sizeHint) {
-                        break;
+                try {
+                    final RecyclerBytesStreamOutput chunkStream = new RecyclerBytesStreamOutput(recycler);
+                    assert target == null;
+                    target = chunkStream;
+                    while (serialization.hasNext()) {
+                        serialization.next().toXContent(builder, params);
+                        if (chunkStream.size() >= sizeHint) {
+                            break;
+                        }
+                    }
+                    if (serialization.hasNext() == false) {
+                        builder.close();
+                    }
+                    final var result = new ReleasableBytesReference(
+                        chunkStream.bytes(),
+                        () -> Releasables.closeExpectNoException(chunkStream)
+                    );
+                    target = null;
+                    return result;
+                } finally {
+                    if (target != null) {
+                        assert false : "failure encoding chunk";
+                        IOUtils.closeWhileHandlingException(target);
+                        target = null;
+                        serialization.forEachRemaining(ignored -> {});
```

Comment:
Why do we need this? At least right now we don't have any resources that have to get released here, so the GC might as well take care of this?

Reply:
Yeah not necessary _today_ indeed, just some extra paranoia. I bet we end up with needing to release something here eventually.

But then again we don't exhaust the iterator if the channel closes so maybe there's more to do in this area anyway.

Reply:
> In particular, if we don't exhaust the iterator here then I worry that we don't return true from future

That shouldn't be the case, the futures should be resolved when running into any exception.

> But then again we don't exhaust the iterator if the channel closes so maybe there's more to do in this area anyway.

Yea, that was by design. So far the iterator should be nowhere near this tricky. I guess it would be nice if we had something where e.g. a bunch of search hits would be iterated through and then released one by one as we iterate and things like that but for now I didn't design for anything like that.